qiskit/primitives/primitive_future.py

```diff
@@ -0,0 +1,54 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Future class for primitives.
+"""
+
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Generic, TypeVar
+
+from .estimator_result import EstimatorResult
+from .sampler_result import SamplerResult
+
+R = TypeVar("R", SamplerResult, EstimatorResult)  # pylint: disable=invalid-name
+
+
+class PrimitiveFuture(ABC, Generic[R]):
```

Comment:
Is this class necessary? Implementors could use the [existing job interface](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/job.py) which already has a very similar interface to this. It should make use of a concurrent.futures.Future internally to implement the asynchronous result once a job (a primitive call) is submitted.

Reply:
Thank you fro your suggestion. I am not sure how to construct when Primitives don't depend on Backend, and what is returned with the `backend` method. Also, I don't think `Job.submit` is necessary. I am also concerned that it will be a circular dependency to `qiskit.provider`. If we have any ideas on how to overcome these issues, it would be nice to use a Job so we could use existing assets.

Reply:
`Job` does use `Backend` in its constructor, but it doesn't have to use it, and if not necessary, you can create one without it `Job(None, job_id, ...)` if there is no backend in the context which would cause `backend()` to return `None`. Would depend on your implementation. In the case of `.submit()`, you might just raise an Unsupported exception or [set the method to None](https://docs.python.org/3/library/exceptions.html#NotImplementedError).

Reply:
In that case, I need to change the job to take `Optional[Backend]`. I will have to ask the Terra core team about this.

Reply:
IBMQJob always [raises an exception](https://github.com/Qiskit/qiskit-ibmq-provider/blob/master/qiskit/providers/ibmq/job/ibmqjob.py#L734-L746) for `job.submit()`, so hopefully most people are used to it.

Reply:
OK. Support `Job` in 047bc5c.